### PR TITLE
DIT-291: Complete Case

### PR DESCRIPTION
**Pull request by ed0906**

### CODE REVIEW

File: app/uk/gov/hmrc/tariffclassificationfrontend/audit/AuditService.scala [hunk 1]
```diff
@@ -26,12 +26,13 @@ import uk.gov.hmrc.tariffclassificationfrontend.models.Case
...
 
 import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
 
 @Singleton
 class AuditService @Inject()(auditConnector: DefaultAuditConnector) {
 
   def auditCaseReleased(c: Case)
-                       (implicit hc: HeaderCarrier): Unit = {
+                       (implicit hc: HeaderCarrier): Future[Unit] = {
```
Comment (gokyo) [inline app/uk/gov/hmrc/tariffclassificationfrontend/audit/AuditService.scala]: Why `Future[Unit]`
The type is `Unit`
See here: https://github.com/hmrc/play-auditing/blob/master/src/main/scala/uk/gov/hmrc/play/audit/http/connector/AuditConnector.scala

File: app/uk/gov/hmrc/tariffclassificationfrontend/audit/AuditService.scala [hunk 2]
```diff
@@ -48,14 +49,22 @@ class AuditService @Inject()(auditConnector: DefaultAuditConnector) {
     sendExplicitAuditEvent(CaseReleased, toJson(c))
   }
 
+  def auditCaseCompleted(c: Case)
+                       (implicit hc: HeaderCarrier): Future[Unit] = {
```
Comment (gokyo) [inline app/uk/gov/hmrc/tariffclassificationfrontend/audit/AuditService.scala]: `Unit`

File: app/uk/gov/hmrc/tariffclassificationfrontend/audit/AuditService.scala [hunk 3]
```diff
@@ -48,14 +49,22 @@ class AuditService @Inject()(auditConnector: DefaultAuditConnector) {
...
+    sendExplicitAuditEvent(CaseReleased, toJson(c))
+  }
+
   private def sendExplicitAuditEvent(auditEventType: String, auditPayload: JsValue)
-                                    (implicit hc : uk.gov.hmrc.http.HeaderCarrier): Unit = {
-    auditConnector.sendExplicitAudit(auditType = auditEventType, detail = auditPayload)
+                                    (implicit hc : uk.gov.hmrc.http.HeaderCarrier): Future[Unit] = {
+    Future {
```
Comment (gokyo) [inline app/uk/gov/hmrc/tariffclassificationfrontend/audit/AuditService.scala]: `Unit`

File: app/uk/gov/hmrc/tariffclassificationfrontend/service/CasesService.scala [hunk 4]
```diff
@@ -27,18 +31,41 @@ import scala.concurrent.ExecutionContext.Implicits.global
...
+          Logger.error(s"Could not create Event for case [${original.reference}] with payload [$event]", throwable)
+      })
+      .map(_ => Unit)
+  }
+
+  def releaseCase(original: Case, queue: Queue, operator: Operator)(implicit hc: HeaderCarrier): Future[Case] = {
+    for {
+      updated <- connector.updateCase(original.copy(status = CaseStatus.OPEN, queueId = Some(queue.id)))
+      _ <- addEvent(original, updated, operator)
+      _ <- auditService.auditCaseReleased(updated)
```
Comment (gokyo) [inline app/uk/gov/hmrc/tariffclassificationfrontend/service/CasesService.scala]: once you convert the type from `Future[Unit]` to `Unit`, this line will become:
```
_ = auditService.auditCaseReleased(updated)
```

File: app/uk/gov/hmrc/tariffclassificationfrontend/service/CasesService.scala [hunk 5]
```diff
@@ -27,18 +31,41 @@ import scala.concurrent.ExecutionContext.Implicits.global
...
+
+    val decisionUpdating: Decision = original.decision
+      .getOrElse(throw new IllegalArgumentException("Cannot Complete a Case without a Decision"))
+      .copy(effectiveStartDate = startDate, effectiveEndDate = endDate)
+    val caseUpdating = original.copy(status = CaseStatus.COMPLETED, decision = Some(decisionUpdating))
+
+    for {
+      updated <- connector.updateCase(caseUpdating)
+      _ <- addEvent(original, updated, operator)
+      _ <- auditService.auditCaseCompleted(updated)
```
Comment (gokyo) [inline app/uk/gov/hmrc/tariffclassificationfrontend/service/CasesService.scala]:
```
_ = auditService.auditCaseCompleted(updated)
```

File: test/unit/uk/gov/hmrc/tariffclassificationfrontend/controllers/CompleteCaseControllerSpec.scala [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+  private val caseWithStatusOPEN = Cases.btiCaseExample.copy(status = CaseStatus.OPEN)
+  private val caseWithStatusCOMPLETED = Cases.btiCaseExample.copy(status = CaseStatus.COMPLETED)
+  private val caseWithoutDecision = Cases.btiCaseExample.copy(status = CaseStatus.OPEN, decision = None)
+
+  private implicit val mat: Materializer = app.materializer
+  private implicit val hc: HeaderCarrier = HeaderCarrier()
+
+  private val controller = new CompleteCaseController(new SuccessfulAuthenticatedAction(operator), casesService, messageApi, appConfig)
+
+  override def beforeEach(): Unit = {
```
Comment (gokyo) [inline test/unit/uk/gov/hmrc/tariffclassificationfrontend/controllers/CompleteCaseControllerSpec.scala]: I believe this is not needed if you do not add anything to the parent implementation.

File: test/unit/uk/gov/hmrc/tariffclassificationfrontend/service/CasesServiceSpec.scala [hunk 7]
```diff
@@ -34,21 +36,22 @@ import scala.concurrent.Future
 
 class CasesServiceSpec extends UnitSpec with MockitoSugar with BeforeAndAfterEach {
 
-  private implicit val hc = HeaderCarrier()
+  private implicit val hc: HeaderCarrier = HeaderCarrier()
   private val manyCases = mock[Seq[Case]]
   private val oneCase = mock[Option[Case]]
   private val queue = mock[Queue]
   private val connector = mock[BindingTariffClassificationConnector]
+  private val audit = mock[AuditService]
```
Comment (gokyo) [inline test/unit/uk/gov/hmrc/tariffclassificationfrontend/service/CasesServiceSpec.scala]: you should assert that the audit events have been created, using Mockito (as I did in the controller)

File: test/unit/uk/gov/hmrc/tariffclassificationfrontend/service/CasesService_CompleteCaseSpec.scala [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.tariffclassificationfrontend.service
+
+import java.time._
+
+import org.mockito.ArgumentCaptor
+import org.mockito.ArgumentMatchers._
+import org.mockito.BDDMockito.{given, _}
```
Comment (gokyo) [inline test/unit/uk/gov/hmrc/tariffclassificationfrontend/service/CasesService_CompleteCaseSpec.scala]: _

File: test/unit/uk/gov/hmrc/tariffclassificationfrontend/service/CasesServiceSpec.scala (hunk 7 above)
Comment (gokyo) [reply test/unit/uk/gov/hmrc/tariffclassificationfrontend/service/CasesServiceSpec.scala]: actually I noticed you created two additional files for this service tests

File: app/uk/gov/hmrc/tariffclassificationfrontend/audit/AuditService.scala (hunk 1 above)
Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/audit/AuditService.scala]: For consistency. All the connectors we use have a `Future` return type, this should be no different. Under the hood it is a thread blocking HTTP request so sticking it in a future means we can parallelise it.
Also improves the Service layer code having it as a future

File: app/uk/gov/hmrc/tariffclassificationfrontend/audit/AuditService.scala (hunk 2 above)
Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/audit/AuditService.scala]: For consistency. All the connectors we use have a `Future` return type, this should be no different. Under the hood it is a thread blocking HTTP request so sticking it in a future means we can parallelise it.
Also improves the Service layer code having it as a future

File: test/unit/uk/gov/hmrc/tariffclassificationfrontend/controllers/CompleteCaseControllerSpec.scala (hunk 6 above)
Comment (ed0906) [reply test/unit/uk/gov/hmrc/tariffclassificationfrontend/controllers/CompleteCaseControllerSpec.scala]: removed

File: app/uk/gov/hmrc/tariffclassificationfrontend/service/CasesService.scala (hunk 5 above)
Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/service/CasesService.scala]: sorted

File: app/uk/gov/hmrc/tariffclassificationfrontend/service/CasesService.scala (hunk 4 above)
Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/service/CasesService.scala]: sorted

File: app/uk/gov/hmrc/tariffclassificationfrontend/audit/AuditService.scala (hunk 1 above)
Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/audit/AuditService.scala]: reverted

**Pull request merged**